cleanlab/dataset.py

```diff
@@ -0,0 +1,150 @@
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# cleanlab is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+
+"""Dataset Module
+Supports dataset-level and class-level automated quality, including finding which classes should
+be merged with other classes in your dataset, which classes should be removed,
+and which classes have the best label accuracy, and which have the worst labels.
+"""
+
+import numpy as np
+from count import estimate_joint, compute_confident_joint
+
+
+def find_similar_classes(
+    joint=None,
+    asymmetric=True,
+    *,
+    labels=None,
+    pred_probs=None,
+    confident_joint=None,
+    multi_label=False,
+):
+    """Ranks all pairs of classes by their similarity and returns the ordered pairs as well as
+    a similarity score between 0 and 1. Use this function to determine which classes in a dataset
+    should be merged.
+
+    The method uses the joint distribution of noisy and true labels to compute ontological issues
+    via the approach published in (Northcutt, Jiang, Chuang (2021) JAIR).
+    paper: https://arxiv.org/abs/1911.00068
+
+    This method works by providing any one (and only one) of the following inputs:
+    1. confident_joint
+    2. joint
+    3. labels and pred_probs
+    but it is not necessary to provide a combination, e.g. the labels and pred_probs and joint.
+
+    For parameter info, see the docstring of `count.estimate_joint`
+
+    Parameters
+    ----------
+    See `count.estimate_joint` docstring.
+
+    asymmetric : bool (default: True)
+      If `asymmetric==True`, includes both pairs (class1, class2) and (class2, class1). Use this for
+      finding "is a" relationships where for example "class1 is a class2".
+      If `asymmetric==False`, the pair (class1, class2) will onyl be returned once and order is
+      arbitrary (internally this is just summing score(class1, class2) + score(class2, class1)).
+
+    Returns
+    -------
+        A tuple of 2 ordered lists: (list_of_pairs_of_classes, likelihood_they_should_be_merged)"""
+
+    def _find_largest_pairs(matrix):
+        """Helper function that returns the sorted (descending) (row, column) index pairs
+        and their values for an arbitrary 2-d numpy matrix."""
+
+        # Find indices of similar classes
+        idx = np.array(list(zip(*np.unravel_index(np.argsort(matrix.ravel()), matrix.shape)))[::-1])
+        # Bring max score between 0.1 and 1.0, otherwise score scales inversely with num_classes ^ 2
+        scaling_factor = 10 ** np.floor(-np.log10(np.max(matrix)))
```

Comment:
Ideally the score has intuitive definition, which I'm not sure this rescaled one does?

Reply:
why not just have the scores simply be this?

P(true_label = i | given_label = j)

That seems very interpretable to me

Reply:
I would think p(given|true) is more intuitive since typically you condition on the latent truth? How did you make that choice? Also why do you like the conditional over the joint? The joint is more expressive albeit the numbers are tiny.

I agree the scaling factor is somewhat arbitrary, it just makes it a good experience since the numbers will always be something readable, but definitely plan to improve it and happy to immediately. Or simply remove it.

Reply:
I thought conditioning on given is more natural because that is what user sees in their data.
So eg. I know that if I see example with given label A, it is untrustworthy.

I'm less sure what I could do with the information that many examples truly should be labeled B but are labeled something else (beyond realizing that my data is under-representing the occurrence of class B).

Re conditional vs joint, I'll leave it up to you. I think both are useful depending on the user's goal: joint is more useful if I want to maximally improve my dataset by fixing A->B mislabels that are overall common. conditional is more useful if I want to compare which given label is most/least trustworthy.

Reply:
I like the idea of implementing the noise matrix version of the scoring (because the number will be normalized and bigger), but the problem is that when `symmetric==True`, it lose its meaning since this i computed by summing the P(x,y) + P(y, x) which doen't make sene if its p(x|y) + P(y|x). Added this as something to think about for 2.1 and moving on with using the joint.

Removed the scaling entirely though, so it will return small joint probability score.